### ISSUE TITLE
ensime fails inside of docker container

### ISSUE THREAD

**Issue by MasseGuillaume** [scalacenter/scastie#250]
To reproduce:

```bash
# make shure to have a clean index
git commit -m "-"

sbt sbtRunner/docker
docker run --network=host scalacenter/scastie-sbt-runner:<git hash>
sbt
server/reStart
```
open http://localhost:9000
autocomplete (ctrl + Space)

Result:
500 timeout

**Comment by MasseGuillaume** [comment]
it would be helpful to setup the logging to the console for ensime to see what is happening in the container.

**Comment by MasseGuillaume** [comment]
```
09:54:45.736 INFO  com.olegych.scastie.sbt.Sbt:169 | [info] Loading project definition from /tmp/scastie1541005692379916767/project
09:54:45.988 INFO  com.olegych.scastie.sbt.Sbt:169 | [info] Loading project definition from /tmp/scastie-ensime6545815263258849600/project
09:54:55.467 INFO  com.olegych.scastie.sbt.Sbt:169 | coursier.ResolutionException: 1 locked
09:54:55.468 INFO  com.olegych.scastie.sbt.Sbt:169 |   /root/.coursier/https/repo1.maven.org/maven2/com/lihaoyi/upickle_2.10/0.4.4/upickle_2.10-0.4.4.jar
09:54:55.469 INFO  com.olegych.scastie.sbt.Sbt:169 | [error] (*:update) coursier.ResolutionException: 1 locked
09:54:55.469 INFO  com.olegych.scastie.sbt.Sbt:169 | [error]   /root/.coursier/https/repo1.maven.org/maven2/com/lihaoyi/upickle_2.10/0.4.4/upickle_2.10-0.4.4.jar
09:54:59.432 INFO  com.olegych.scastie.sbt.Sbt:169 | [info] Set current project to scastie-ensime6545815263258849600 (in build file:/tmp/scastie-ensime6545815263258849600/)
09:54:59.552 INFO  com.olegych.scastie.sbt.Sbt:169 | sifOme0Nae
09:54:59.555 INFO  com.olegych.scastie.sbt.EnsimeActor:169 | Generating ensime config file
09:54:59.575 INFO  com.olegych.scastie.sbt.Sbt:169 | [info] ENSIME update.
```

**Comment by rorygraves** [comment]
@dimart  Helpful commands for debugging:
```
docker ps
docker exec -it <dockerInstanceID from ps>  bash
```
This allows you to run a shell within the running container and look at logs etc.

**Comment by rorygraves** [comment]
Try going to the directory that ensime uses to generate the config and then you can do ```sbt ensimeConfig``` directly.

**Comment by MasseGuillaume** [comment]
@rorygraves any update on this? You mentioned that we need an ensime user.

**Comment by rorygraves** [comment]
I'm getting close to having this fully working. 
Just limited amounts of time

**Comment by MasseGuillaume** [comment]
@rorygraves awesome, is there something I can do? Can you push what you have so far?

**Comment by rorygraves** [comment]
Apologies for the wall of text.

tldr; The code is available on the below branch, its currently broken in docker as it interacts badly with the latest ```Sbt``` instance changes in master.  also contains a bunch of changes around sbtActor management

-----

So the latest code is here: https://github.com/rorygraves/scastie/tree/ensimeSupport  (I've just rebased it against master.

Its currently broken after rebasing against @dimart  latest code.  This is because the code has seperated out ensime into its own directory and area.  In docker this is (esime:ensime) under /home/ensime.    However @dimart change means it reuses the existing sbt instance and now I have no idea where it is writing the .ensime/ stuff to - hence the failure to start the process.
One of the reasons for the original failure is that coursier does not handling locking well with multiple sbt instances running (it attempts to lock and fails if it can't immediately acquire the lock).  

I think it makes sense to have a separate area - as it means you could potentially do a build an complete for separate projects at the same time. 

I've made a whole slew of unrelated changes as I was seeing random hangs and failures between the sbtActor and the webserver (probably similar to some we see in production.  So I started adding extra logging and ids to the various client requests to make tracing easier.  
One of the key changes is to have an SbtActorProxy for each remote connection - this manages the connection/disconnection state of each remote engine.    One of the things I noticed was that if a docker instance is lost it will never try and reconnect, and simply remove it from the balancer.  Each SbtActorProxy is now responsible for ping and connecting to its server and will reconnect after loss.
This is very noisy however - I see dead letters and lots of connection information noise when the sbtActor is down.   I tried to suppress this by moving the SbtPing message to a package where I can subclass DeadLetterSuppression, but I am not sure I have understood the behaviour.

I'm not going to have time to work on this until the weekend now - there are several things.
 to be done.

1. The seperate user might not be necessary if ensime is using the same Sbt instance, although some messing arround may be needed to fix it under docker.
2. We could easily env flag it in (see the way ENSIME_USER_GROUP is defined) to enable/disable Ensime rather than having to change the code in SbtMain
3. Move tasks into a queue within the load balancer rather than relying on the queuing in the sbtActor - means we can handle failure more gracefully.
4. Update the status page to show more information (which sbtActors are up and down and stats info, rather than relying on the stats from the balancer which only shows 'active engines'
5. Update the config to supports sbtActors from multiple machines.

Future suggestions:
1. Move away from Akka remoting - use websockets which are sbtActor driven (it connects to the server rather than the other way round).  This also removes the need for (5) and gives a cleaner behaviour.

**Comment by MasseGuillaume** [comment]
@rorygraves @dimart 

@rorygraves I will take a look at this today, I hope I can unblock us.

**Comment by MasseGuillaume** [comment]
I got it working. The trick is to run under the ensime user. I ran the sbtRunner inside the docker container and got the autocompletion working.

**Comment by rorygraves** [comment]
Running the everything under the ensime user?

**Comment by MasseGuillaume** [comment]
@rorygraves exactly. I'm building up a PR, should be ready soon.

**Comment by dimart** [comment]
@rorygraves @MasseGuillaume thank you so much for your help!

**Comment by MasseGuillaume** [comment]
@dimart I also did #266 and #267

**Comment by rorygraves** [comment]
I think sharing the sbt instance had fixed the original issue with coursier/ivy2
